### PR TITLE
fix(backend)!: derive BTC pending-tx address from caller principal

### PR DESCRIPTION
**Pull request by sbpublic**
# Motivation

Fix HackenProof report ICPBB-120: `btc_get_pending_transactions` accepted a caller-supplied address and used the UTXOs returned for it as the input to a prune pass over **all** of the caller's pending BTC transactions. Passing a well-formed-but-foreign address (zero UTXOs) wiped every non-expired reservation across all the caller's stored addresses, defeating the UTXO double-reservation guard in `btc_add_pending_transaction`. Severity is Low — Bitcoin consensus still prevents the actual double-spend — but the self-affecting UX guard is undermined.

# Changes

The endpoint now mirrors `btc_add_pending_transaction` and derives the P2WPKH address from the caller's principal via `signer::btc_principal_to_p2wpkh_address`. `params.address` is no longer read anywhere, so the field is removed from `BtcGetPendingTransactionsRequest` and the chain of now-dead helpers (`validate_address`, `MAX_ADDRESS_LEN`, the `Validate` / `validate_on_deserialize!` impls for the request, the matching `test_validate_on_deserialize!` block, and the integration-test `MOCK_ADDRESS` constant) is removed.

Defense in depth in the prune logic: `prune_pending_transactions` now treats an empty `current_utxos` slice as "no confirmation information available" rather than "all utxos confirmed", so a caller fetching utxos for an address with none can no longer wipe their non-expired reservations. The 1-hour age rule still fires unchanged, so genuinely stale reservations are still released — just on the time budget alone for the all-spent-to-zero edge case.

Frontend ripple: `BtcGetPendingTransactionParams` (and its inherited `BtcAddPendingTransactionParams`) drops `address`, every call site stops sending it, and the matching FE tests are updated. The store-keying `address` local in `btc-pending-sent-transactions.services.ts` is preserved — only the backend argument drops it. `BtcWalletScheduler.loadBtcPendingTransactionsData` also drops the now-unused `btcAddress` parameter from its private signature.

`PRODUCT.md` gains a Bitcoin section recording the pending-transaction reservation guarantee and the explicit negative guarantee that the caller cannot specify which address's pending transactions are read, added, or pruned.

Note on generated bindings: `src/backend/backend.did` and the files under `src/declarations/backend/` were hand-edited (4 surgical removals of the `address` field) because the local toolchain cannot cross-compile `secp256k1-sys` to `wasm32-unknown-unknown`, so `npm run generate` could not run end-to-end here. CI's `binding-checks` job ran a full regen on a wasm-capable runner and the result matches these edits.

# Tests

Two new backend unit tests in `pending_tx_model.rs` pin the prune-guard behaviour:

- `test_empty_current_utxos_retains_non_expired` — seeds a non-expired pending tx and prunes with `&[]`; asserts the tx is retained. Fails against `main`.
- `test_empty_current_utxos_still_prunes_expired` — seeds an expired pending tx and prunes with `&[]`; asserts the tx is removed (age rule still applies).

Backend integration tests in `tests/it/bitcoin.rs` updated to construct `BtcGetPendingTransactionsRequest` without `address`. The compiler now enforces that no caller can supply an address, replacing the previous runtime check.

Frontend test updates: `backend.api.spec.ts`, `backend.canister.spec.ts`, `btc-send.services.spec.ts`, `btc-open-crypto-pay.services.spec.ts`.

Local gates run: backend native clippy clean, backend `pending_tx_model` unit tests 19/19 (including the two new ones), backend integration tests compile cleanly, `lint.did.sh` clean, vitest 222/222 on touched FE files. Wasm clippy, the full pocket-ic suite, and the full ESLint pass are deferred to CI (local toolchain limitations).

BREAKING CHANGE: `BtcGetPendingTransactionsRequest` no longer carries an `address` field. The endpoint derives the BTC address from the caller's principal — there is no way to read pending transactions for any other address. The FE bindings under `src/declarations/backend/` are updated in this PR; external consumers must regenerate or hand-update their bindings.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude Opus 4.7 (`claude-opus-4-7`)

### CODE REVIEW

File: src/backend/src/api/bitcoin.rs [hunk 1]
```diff
@@ -184,9 +184,13 @@ pub async fn btc_get_pending_transactions(
         )
         .map_err(|msg| BtcGetPendingTransactionsError::InvalidDelegationChain { msg })?;
 
+        let source_address = signer::btc_principal_to_p2wpkh_address(params.network, &principal)
+            .await
+            .map_err(|msg| BtcGetPendingTransactionsError::InternalError { msg })?;
```
Comment (AntonioVentilii) [inline src/backend/src/api/bitcoin.rs:189]: the backend now recalculates the user bitcoin address on every call instead of trusting the one the app sends.

This recalculation is an extra network call that's never cached, and this endpoint is polled every 30 seconds per user (worker loop).

So, all correct and well, but just to be aware that it adds a small repeated cost to a busy loop

Ideally, we should cache it, but it can be for another iteration

**Review by AntonioVentilii — APPROVED** [review]
LGTM tks, with a small non-blocker obs

File: src/backend/src/api/bitcoin.rs (hunk 1 above)
Comment (sbpublic) [reply src/backend/src/api/bitcoin.rs:189]: Good catch — confirming the trade-off.

**Summary**: `btc_get_pending_transactions` now derives the source address via `signer::btc_principal_to_p2wpkh_address` instead of trusting `params.address`. That derive is an inter-canister call to the chain-fusion signer, executed on every endpoint call — including the 30-second worker poll.

**Analysis**: the derive is deterministic for a given principal (P2WPKH from the threshold-ECDSA pubkey), so a per-principal cache is straightforward and would eliminate the recurring cost. Out of scope for the security fix here; better as its own PR so cache lifecycle, key-rotation invariants, and reuse by `btc_add_pending_transaction` (and any future BTC endpoint that derives the same address) can be designed together.

**Severity**: low (perf observation, no correctness issue).

**Fix**: no code change in this PR. Spawning a follow-up to cache the derived P2WPKH address per principal so all BTC endpoints share one derive.

🤖 Claude Opus 4.7

File: src/declarations/backend/backend.did [hunk 2]
```diff
@@ -327,8 +327,7 @@ type BtcGetPendingTransactionsReponse = record {
 };
 type BtcGetPendingTransactionsRequest = record {
 	ii_delegation_chain : opt IIDelegationChain;
-	network : Network;
-	address : text
+	network : Network
 };
```
[the remark this answers is not in the document]
Comment (sbpublic) [reply src/declarations/backend/backend.did:336]: Good catch — confirming the empirical check.

**Summary**: `backend.did`, `backend.factory.did.js`, and `backend.factory.certified.did.js` under `src/declarations/backend/` (plus `backend.did.d.ts`) were hand-edited to mirror the source-of-truth change in `src/shared/src/types/bitcoin.rs`. The local toolchain couldn't cross-compile `secp256k1-sys` to `wasm32-unknown-unknown`, so `npm run generate` couldn't run end-to-end on this branch.

**Analysis**: CI's `binding-checks` job ran a full `npm run generate` on a wasm-capable runner against this branch and went green, which means the hand-edited output already matches a clean regen byte-for-byte. The drift risk you're flagging is already gated by CI here. For future similar PRs the cleaner alternative is `gh run download` of the CI artifact rather than hand-editing — noted.

**Severity**: low (general code-hygiene concern; empirical drift is gated by CI here).

**Fix**: no change in this PR. Also resolves your related comments on `backend.factory.did.js` and `backend.factory.certified.did.js` — same root cause.

🤖 Claude Opus 4.7

File: src/declarations/backend/backend.factory.certified.did.js [hunk 3]
```diff
@@ -166,8 +166,7 @@ export const idlFactory = ({ IDL }) => {
 	});
 	const BtcGetPendingTransactionsRequest = IDL.Record({
 		ii_delegation_chain: IDL.Opt(IIDelegationChain),
-		network: Network,
-		address: IDL.Text
+		network: Network
 	});
```
[the remark this answers is not in the document]
Comment (sbpublic) [reply src/declarations/backend/backend.factory.certified.did.js:170]: Good catch — same as the `backend.did` thread.

**Summary**: identical Copilot finding on a sibling generated file, addressed at https://github.com/dfinity/oisy-wallet/pull/13060#discussion_r3403756910.

**Analysis**: same root cause — CI's `binding-checks` regenerated this file on a wasm-capable runner and went green, so the hand edit matches the canonical regen byte-for-byte.

**Severity**: low (same as the sibling thread).

**Fix**: no change — see the resolution on the `backend.did` thread.

🤖 Claude Opus 4.7

File: src/declarations/backend/backend.factory.did.js [hunk 4]
```diff
@@ -166,8 +166,7 @@ export const idlFactory = ({ IDL }) => {
 	});
 	const BtcGetPendingTransactionsRequest = IDL.Record({
 		ii_delegation_chain: IDL.Opt(IIDelegationChain),
-		network: Network,
-		address: IDL.Text
+		network: Network
 	});
```
[the remark this answers is not in the document]
Comment (sbpublic) [reply src/declarations/backend/backend.factory.did.js:170]: Good catch — same as the `backend.did` thread.

**Summary**: identical Copilot finding on a sibling generated file, addressed at https://github.com/dfinity/oisy-wallet/pull/13060#discussion_r3403756910.

**Analysis**: same root cause — CI's `binding-checks` regenerated this file on a wasm-capable runner and went green, so the hand edit matches the canonical regen byte-for-byte.

**Severity**: low (same as the sibling thread).

**Fix**: no change — see the resolution on the `backend.did` thread.

🤖 Claude Opus 4.7
